### PR TITLE
feat: mnlistdiff v20 CL sig quorums 

### PR DESCRIPTION
**Pull request by ogabrielides**
## Issue being fixed or feature implemented
Implementation of Randomness Beacon Part 3.

Starting from v20 activation fork, members for quorums are sorted using (if available) the best CL signature found in Coinbase.
If no CL signature is present yet, then the usual way is used (By using Blockhash instead)

The actual new way to shuffle is already implemented in https://github.com/dashpay/dash/pull/5366.

SPV clients also need to calculate members, but they only know block headers. 
Since Coinbase is in the actual block, then they lack the required information to correctly calculate quorum members.

## What was done?
-  Message `MNLISTIDFF` is enriched with a new field `quorumsCLSigs`. This field holds the Chainlock Signature required for each set of indexes corresponding to quorums in field `newQuorums`.
-  Protocol version has been bumped to `70230`.
-  Clients with protocol version greater or equal to `70230` will receive the new field `quorumsCLSigs`.
- The same field is returned in `protx diff` RPC.

Note:
- Field `quorumsCLSigs` will populated only after v20 activation
- If for one or more quorums, no non-null CL sig was found in CbTx then a null signature is returned in `quorumsCLSigs`.

## How Has This Been Tested?
- Functional test mininode's protocol version was bumped to `70230`.
- `feature_llmq_rotation.py` checks that `quorumsCLSigs` match in both P2P and RPC messages.

## Breaking Changes
No

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

**Comment by ogabrielides** [comment]
@UdjinM6 Why CI isn't starting?

File: src/evo/simplifiedmns.cpp [hunk 1]
```diff
@@ -178,6 +180,66 @@ bool CSimplifiedMNListDiff::BuildQuorumsDiff(const CBlockIndex* baseBlockIndex,
...
+            // In case of non-rotation, quorums rely on the CL sig expected in the block of the DKG - 8
+            pWorkBaseBlockIndex = blockIndex->GetAncestor(q->m_quorum_base_block_index->nHeight - 8);
+        }
+
+        workBaseBlockIndexMap.insert(std::make_pair(pWorkBaseBlockIndex, idx));
+        idx++;
+    }
+
+    auto it = workBaseBlockIndexMap.begin();
+    while (it != workBaseBlockIndexMap.end()) {
```
Comment (knst) [inline src/evo/simplifiedmns.cpp:214]: nit:
it looks very much as `for` loop for me:
for (auto it  = workBaseBlockIndexMap.begin(); it != workBaseBlockIndexMap.end(); it = range.second;) {...}

File: src/evo/simplifiedmns.cpp [hunk 2]
```diff
@@ -178,6 +180,66 @@ bool CSimplifiedMNListDiff::BuildQuorumsDiff(const CBlockIndex* baseBlockIndex,
...
+        // Iterate the std::multimap once per key (CBlockIndex containing the expected CL signature in CbTx)
+        const CBlockIndex* pWorkBaseBlockIndex = it->first;
+        auto cbcl = GetNonNullCoinbaseChainlock(pWorkBaseBlockIndex);
```
Comment (knst) [inline src/evo/simplifiedmns.cpp:217]: nit: `cbcl` is not changing - better to have a const modifier here

File: src/evo/simplifiedmns.cpp [hunk 3]
```diff
@@ -178,6 +180,66 @@ bool CSimplifiedMNListDiff::BuildQuorumsDiff(const CBlockIndex* baseBlockIndex,
...
+        CBLSSignature sig;
+        if (cbcl.has_value()) {
+            sig = cbcl.value().first;
+        }
+        // Get the range of indexes (values) for the current key and merge them into a single std::set
+        auto range = workBaseBlockIndexMap.equal_range(pWorkBaseBlockIndex);
```
Comment (knst) [inline src/evo/simplifiedmns.cpp:223]: I think better to have a `const` modifier here for range - the `range` is not changing anyhow

File: src/evo/simplifiedmns.h [hunk 4]
```diff
@@ -135,6 +144,10 @@ class CSimplifiedMNListDiff
     std::vector<std::pair<uint8_t, uint256>> deletedQuorums; // p<LLMQType, quorumHash>
     std::vector<llmq::CFinalCommitment> newQuorums;
 
+    // Map of Chainlock Signature used for shuffling per set of quorums
+    // The set of quorums is the set of indexes corresponding to entries in newQuorums
+    std::map<CBLSSignature, std::set<uint16_t>, BLSSignatureCompare> quorumsCLSigs;
```
Comment (knst) [inline src/evo/simplifiedmns.h:149]: nit: why not lambda here?

Comment (ogabrielides) [reply src/evo/simplifiedmns.h:149]: @knst I don't think a lamda would make the code cleaner necessary

**Review by thephez — COMMENTED** [review]
A few minor suggestions

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by UdjinM6 — DISMISSED** [review]
re-utACK

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]
Please see f838f959f437fd41e8b826fce87d0cff0c1cc246 c02bac53226131c97d0d219c84fe36fc6a945824 and 63efe7f0a00ba7240704154211c9bfc1b2c6d1d2

Then, I am concerned about testing for this, I tried to break the test but couldn't really. I think any of these changes should break the tests

00f1a42cbe4d88f4cd0b67056568f03a6312633b
7f3f385540e22396a2908df9401addd93c08a90f
5c04684ebc58008135e0eaf8c86db767333fb6c9

**Comment by ogabrielides** [comment]
@PastaPastaPasta Implemented your suggestions.

Regarding the tests, currently `feature_llmq_rotation.py` checks that `quorumsCLSigs` match in both P2P and RPC messages.

There is no way to check that those CLSigs are indeed used for shuffling in members.

I will think about that, but any ideas are welcomed.
@UdjinM6

**Comment by ogabrielides** [comment]
@PastaPastaPasta @UdjinM6 
Implemented quorum verification using CLSigs returned in protx diff.
Somes notes:
- By `quorum verification` I mean only checking that the calculated set of members using the CLSig is the same with the actual members of the quorum. No BLS verification on the quorum's public key. (Could be done in the future)
- null CLSigs are skipped: No need to verify old way of shuffling (using BlockHash)
- Rotated quorums are skipped: Too complicated to implement. Will be done in a separate PR once this is merged.

The commits https://github.com/dashpay/dash/commit/00f1a42cbe4d88f4cd0b67056568f03a6312633b https://github.com/dashpay/dash/commit/7f3f385540e22396a2908df9401addd93c08a90f https://github.com/dashpay/dash/commit/5c04684ebc58008135e0eaf8c86db767333fb6c9 are still not making the test to fail. These commits are basically disturbing the workBlockIndex -> resulting into selecting the wrong CL from `CbTx`.

Although, in functional tests, mining is done so fast, that CL isn't updated very often in `CbTx` since we didn't have time to get ChainLocked.

Basically we need to include more `wait_for_chainlocked_block_all_nodes` in the test to make those commit fail.

**Comment by ogabrielides** [comment]
@PastaPastaPasta @UdjinM6 

With the latest pushed change, quorum verification captures the disturbed returned CL caused by https://github.com/dashpay/dash/commit/7f3f385540e22396a2908df9401addd93c08a90f https://github.com/dashpay/dash/commit/5c04684ebc58008135e0eaf8c86db767333fb6c9.

File: src/version.h [hunk 5]
```diff
@@ -9,9 +9,7 @@
 /**
  * network protocol versioning
  */
-
-
-static const int PROTOCOL_VERSION = 70229;
+static const int PROTOCOL_VERSION = 70230;
```
Comment (PastaPastaPasta) [inline src/version.h:12]: please also bump 
```
static const int MIN_MASTERNODE_PROTO_VERSION = 70227;
```

Comment (UdjinM6) [reply src/version.h:12]: and pls bring back empty lines you deleted here

**Review by UdjinM6 — APPROVED** [review]
utACK
